src/v/serde/avro/parser.cc

```diff
@@ -0,0 +1,202 @@
+            co_return std::make_unique<parsed::message>(_parser.read_bool());
+        case ::avro::AVRO_NULL:
+            co_return std::make_unique<parsed::message>(
+              parsed::primitive{parsed::avro_null{}});
+        case ::avro::AVRO_RECORD: {
+            auto record = std::make_unique<parsed::record>();
+
+            for (auto i : boost::irange<size_t>(node->leaves())) {
+                auto field_node = node->leafAt(i);
+                record->fields.push_back(
+                  co_await parse_node(field_node, state));
+            }
+            co_return std::make_unique<parsed::message>(std::move(record));
+        }
+        case ::avro::AVRO_ARRAY: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::list>();
+
+            if (element_count > 0) {
+                ret->elements.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    ret->elements.push_back(
+                      co_await parse_node(node->leafAt(0), state));
+                }
+                // avro always writes 0 as the array footer, first element count
+                // field is skipped if array is empty
+                _parser.read_varlong();
+            }
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_MAP: {
+            auto element_count = decode_element_count();
+            auto ret = std::make_unique<parsed::map>();
+            if (element_count > 0) {
+                ret->entries.reserve(element_count);
+                for (size_t i = 0; i < element_count; ++i) {
+                    auto key = parse_string();
+                    ret->entries.emplace_back(
+                      std::move(key),
+                      co_await parse_node(node->leafAt(1), state));
+                }
+                // avro always writes 0 as the map footer, first element count
+                // field is skipped if the map is empty
+                _parser.read_varlong();
+            }
+            co_return std::make_unique<parsed::message>(std::move(ret));
+        }
+        case ::avro::AVRO_UNION: {
+            auto [idx, _] = _parser.read_varlong();
+            if (idx < 0 || idx >= static_cast<int64_t>(node->leaves())) {
+                throw std::invalid_argument(fmt::format(
+                  "Invalid union branch decoded, it must be greater than 0 and "
+                  "smaller than number of leaves in union {}. current: {}",
+                  node->leaves(),
+                  idx));
+            }
+
+            auto union_branch = node->leafAt(idx);
+            co_return std::make_unique<parsed::message>(
+              std::make_unique<parsed::avro_union>(parsed::avro_union{
+                .branch = static_cast<size_t>(idx),
+                .message = co_await parse_node(union_branch, state)}));
+        }
+        case ::avro::AVRO_FIXED: {
```

Comment:
I was thinking to deferring logical type understanding to the caller. I think it would be easier and we wouldn't have to proliferate the `parsed` type hierarchy